### PR TITLE
feat: array functions treat an array as an element

### PR DESCRIPTION
**Pull request by izveigor**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/6985
Follow on to https://github.com/apache/arrow-datafusion/pull/6879
Follow on to https://github.com/apache/arrow-datafusion/pull/6805

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?
Yes

# Are there any user-facing changes?
Yes

### CODE REVIEW

**Comment by alamb** [comment]
Thank you for this PR @izveigor  -- I ran out of time today but I have this PR on my list to review tomorrow

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 1]
```diff
@@ -554,28 +595,36 @@ select array_concat(column1, column2) from arrays_values_v2;
 [11, 12]
 NULL
 
-# TODO: Concat columns with different dimensions fails
-# array_concat column-wise #10 (1D + 2D)
-# query error DataFusion error: Arrow error: Invalid argument error: column types must match schema types, expected List\(Field \{ name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\) but found List\(Field \{ name: "item", data_type: List\(Field \{ name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\) at column index 0
-# select array_concat(make_array(column3), column4) from arrays_values_v2;
+# array_concat column-wise #9 (2D + 1D)
+query ?
+select array_concat(column4, make_array(column3)) from arrays_values_v2;
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:600]: Does 1D + 2D pass?

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 2]
```diff
@@ -554,28 +595,36 @@ select array_concat(column1, column2) from arrays_values_v2;
...
+[[]]
+
+# array_concat column-wise #10 (3D + 2D + 1D)
+query ?
+select array_concat(column4, column1, column2) from nested_arrays;
+----
+[[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]], [[1, 2, 3], [2, 9, 1], [7, 8, 9], [1, 2, 3], [1, 7, 4], [4, 5, 6]], [[7, 8, 9]]]
+[[[11, 12, 13], [14, 15, 16]], [[17, 18, 19], [20, 21, 22]], [[4, 5, 6], [10, 11, 12], [4, 9, 8], [7, 8, 9], [10, 11, 12], [1, 8, 7]], [[10, 11, 12]]]
 
-# array_concat column-wise #11 (1D + Integers)
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:562]: I think we don't need to support this.

File: datafusion/physical-expr/src/array_expressions.rs [hunk 3]
```diff
@@ -543,7 +545,18 @@ fn align_array_dimensions(args: Vec<ArrayRef>) -> Result<Vec<ArrayRef>> {
...
                 for _ in 0..(max_ndim - ndim) {
                     let data_type = aligned_array.as_ref().data_type().clone();
-                    aligned_array = array_array(&[aligned_array], data_type)?;
+                    let offsets: Vec<i32> =
+                        (0..downcast_arg!(aligned_array, ListArray).offsets().len())
+                            .map(|i| i as i32)
+                            .collect();
+                    let field = Arc::new(Field::new("item", data_type, true));
+
+                    aligned_array = Arc::new(ListArray::try_new(
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/physical-expr/src/array_expressions.rs:554]: There are `downcast_arg()` in this file, `as_list_array` in `datafusion/common/src/cast.rs` and `as_list, as_list_opt` in arrow-array. Which one to use among these methods?

**Review by jayzhan211 — APPROVED** [review]
LGTM, thanks!

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 4]
```diff
@@ -55,6 +55,13 @@ AS VALUES
   (make_array(make_array(15, 16),make_array(NULL, 18)), make_array(16.6, 17.7, 18.8), NULL)
 ;
 
+statement ok
+CREATE TABLE nested_arrays
+AS VALUES
+  (make_array(make_array(1, 2, 3), make_array(2, 9, 1), make_array(7, 8, 9), make_array(1, 2, 3), make_array(1, 7, 4), make_array(4, 5, 6)), make_array(7, 8, 9), 2, make_array([[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]])),
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:61]: I recommend adding NULL values -- both as elements as well as the arrays as that is often a corner case that gets overlooked

Perhaps something like 

```suggestion
  (make_array(make_array(1, 2, 3), NULL, make_array(7, 8, 9), make_array(1, 2, 3), make_array(1, 7, 4), make_array(4, 5, 6)), make_array(7, 8, 9), 2, make_array([[NULL, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]])),
```

File: datafusion/physical-expr/src/array_expressions.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/array_expressions.rs:554]: The last argument to `try_new` is the `Option<NullBuffer>` right which defines which elements are nulls.

Perhaps I don't understand your question

**Review by alamb — APPROVED** [review]
Thank you @izveigor  -- the tests are especially nice.

I think adding some coverage of nulls would improve things but this PR is an improvement over master so 👍

**Comment by alamb** [comment]
Thanks @izveigor  and @jayzhan211

**Pull request merged**